scripts/push_sdk/tests/test_push_php_sdk.py

```diff
@@ -0,0 +1,183 @@
+    
+    # Act & Assert
+    with pytest.raises(InvalidCriteoServiceException):
+      pipeline.execute()
+  
+  def test_execute_should_fail_when_invalid_criteo_service(self):
+    # Arrange
+    invalid_criteo_service = 'invalid_criteo_service'
+    
+    fs_client = self.fs_client_builder.that_responds_on_list_dir(self.generated_sources, [f'{invalid_criteo_service}_preview']).client
+    pipeline = PushPhpSdkAction(self.git_client_builder.client, fs_client, self.os_client_builder.client)
+    
+    # Act & Assert
+    with pytest.raises(InvalidCriteoServiceException):
+      pipeline.execute()
+      
+  def test_execute_should_fail_when_invalid_api_version(self):
+    # Arrange
+    invalid_api_version = 'invalid-api-version'
+    
+    fs_client = self.fs_client_builder.that_responds_on_list_dir(self.generated_sources, [f'{CriteoService.marketingsolutions}_{invalid_api_version}']).client
+    pipeline = PushPhpSdkAction(self.git_client_builder.client, fs_client, self.os_client_builder.client)
+    
+    # Act & Assert
+    with pytest.raises(InvalidApiVersionException):
+      pipeline.execute()
+  
+  def test_execute_should_fail_when_remove_previous_sources_fails(self):
+    # Arrange
+    expected_exception = FileNotFoundError()
+    
+    fs_client = self.fs_client_builder.that_fails_on_remove(expected_exception).client
+    pipeline = PushPhpSdkAction(self.git_client_builder.client, fs_client, self.os_client_builder.client)
+    
+    # Act & Assert
+    with pytest.raises(Exception):
+      pipeline.execute()
+  
+  def test_execute_should_fail_when_copy_new_sources_fails(self):
+    # Arrange
+    expected_exception = FileNotFoundError()
+    
+    fs_client = self.fs_client_builder.that_fails_on_copy(expected_exception).client
+    pipeline = PushPhpSdkAction(self.git_client_builder.client, fs_client, self.os_client_builder.client)
+    
+    # Act & Assert
+    with pytest.raises(Exception):
+      pipeline.execute()
+      
+  def test_execute_should_fail_when_copy_new_sources_fails(self):
+    # Arrange
+    expected_exception = FileNotFoundError()
+    
+    fs_client = self.fs_client_builder.that_fails_on_copy(expected_exception).client
+    pipeline = PushPhpSdkAction(self.git_client_builder.client, fs_client, self.os_client_builder.client)
+    
+    # Act & Assert
+    with pytest.raises(Exception):
+      pipeline.execute()
+  
+  test_data = [
+    ('preview', 0, f'0.0.{get_formatted_date()}'),
+    ('preview', 1, f'0.0.{get_formatted_date()}-patch1'),
+    ('preview', 99, f'0.0.{get_formatted_date()}-patch99'),
+    ('2021-10', 0, f'2021.10.0.{get_formatted_date()}'),
+    ('2021-10', 1, f'2021.10.0.{get_formatted_date()}-patch1'),
+    ('2021-10', 99, f'2021.10.0.{get_formatted_date()}-patch99'),
+  ]
+
+  @pytest.mark.parametrize("api_version, failure_count, expected_tag", test_data)
+  def test_tag_successful(self, api_version, failure_count, expected_tag):
```

Comment:
Also curious, what happens in the off chance that the test begins running at 11:59:59, so that the get_formmatted_date is different from the execution vs the test_data? Probably a very unlikely scenario, but would this test handle that?

Reply:
you're right, I've updated the `utils.get_formatted_date` so that it generates the date only once, and reuse it when the method is called again, a bit like a singleton